File: ethstats/src/main/java/org/apache/tuweni/ethstats/EthStatsReporter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
 * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
 * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
 * License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package org.apache.tuweni.ethstats;

import org.apache.tuweni.concurrent.AsyncCompletion;
import org.apache.tuweni.concurrent.CompletableAsyncCompletion;
import org.apache.tuweni.eth.EthJsonModule;
import org.apache.tuweni.units.bigints.UInt256;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URI;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;
import java.util.function.Supplier;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import io.vertx.core.MultiMap;
import io.vertx.core.Promise;
import io.vertx.core.TimeoutStream;
import io.vertx.core.Vertx;
import io.vertx.core.WorkerExecutor;
import io.vertx.core.http.HttpClient;
import io.vertx.core.http.HttpClientOptions;
import io.vertx.core.http.WebSocket;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * ETHNetStats reporting service.
 * <p>
 * This service connects to a running ethnetstats service and reports.
 * <p>
 * If the service is not available, the reporter will keep trying to connect periodically. The service will report
 * statistics over time.
 */
public final class EthStatsReporter {

  private final static ObjectMapper mapper = new ObjectMapper();
  private final static Logger logger = LoggerFactory.getLogger(EthStatsReporter.class);

  static {
    mapper.registerModule(new EthJsonModule());
  }

  private final static long DELAY = 5000;
  private final static long REPORTING_PERIOD = 1000;
  private final static long PING_PERIOD = 15000;


  private final String id;
  private final Vertx vertx;
  private final List<URI> ethstatsServerURIs;
  private final AtomicBoolean started = new AtomicBoolean(false);
  private final AtomicBoolean waitingOnPong = new AtomicBoolean(false);
  private final NodeInfo nodeInfo;
  private final String secret;
  private final AtomicReference<Integer> newTxCount = new AtomicReference<>();
  private final Consumer<List<UInt256>> historyRequester;
  private final Supplier<Long> timeSupplier;

  private WorkerExecutor executor;
  private HttpClient client;
  private AtomicReference<BlockStats> newHead = new AtomicReference<>();
  private AtomicReference<NodeStats> newNodeStats = new AtomicReference<>();
  private AtomicReference<List<BlockStats>> newHistory = new AtomicReference<>();

  /**
   * Default constructor.
   *
   * @param vertx a Vert.x instance, externally managed.
   * @param id the id of the ethstats reporter for communications
   * @param ethstatsServerURIs the URIs to connect to eth-netstats, such as ws://www.ethnetstats.org:3000/api. URIs are
   *        tried in sequence, and the first one to work is used.
   * @param secret the secret to use when we connect to eth-netstats
   * @param name the name of the node to be reported in the UI
   * @param node the node name to be reported in the UI
   * @param port the devp2p port exposed by this node
   * @param network the network id
   * @param protocol the version of the devp2p eth subprotocol, such as eth/63
   * @param os the operating system on which the node runs
   * @param osVer the version of the OS on which the node runs
   * @param historyRequester a hook for ethstats to request block information by number.
   * @param timeSupplier a function supplying time in milliseconds since epoch.
   */
  public EthStatsReporter(
      Vertx vertx,
      String id,
      List<URI> ethstatsServerURIs,
      String secret,
      String name,
      String node,
      int port,
      String network,
      String protocol,
      String os,
      String osVer,
      Consumer<List<UInt256>> historyRequester,
      Supplier<Long> timeSupplier) {
    this.id = id;
    this.vertx = vertx;
    this.ethstatsServerURIs = ethstatsServerURIs;
    this.secret = secret;
    this.nodeInfo = new NodeInfo(name, node, port, network, protocol, os, osVer);
    this.historyRequester = historyRequester;
    this.timeSupplier = timeSupplier;
  }

  public AsyncCompletion start() {
    if (started.compareAndSet(false, true)) {
      executor = vertx.createSharedWorkerExecutor("ethnetstats");
      client = vertx.createHttpClient(new HttpClientOptions().setLogActivity(true));
      CompletableAsyncCompletion completion = AsyncCompletion.incomplete();
      startInternal(completion);
      return completion;
    }
    return AsyncCompletion.COMPLETED;
  }

  public void stop() {
    if (started.compareAndSet(true, false)) {
      logger.debug("Stopping the service");
      executor.close();
    }
  }

  public void sendNewHead(BlockStats newBlockStats) {
    newHead.set(newBlockStats);
  }

  public void sendNewPendingTransactionCount(int txCount) {
    newTxCount.set(txCount);
  }

  public void sendNewNodeStats(NodeStats nodeStats) {
    newNodeStats.set(nodeStats);
  }

  public void sendHistoryResponse(List<BlockStats> blocks) {
    newHistory.set(blocks);
  }

  private void startInternal(CompletableAsyncCompletion completion) {
    for (URI uri : ethstatsServerURIs) {
      executor.executeBlocking((Promise<Boolean> handler) -> connect(handler, uri), result -> {
        logger.debug("Attempting to connect", result.cause());
        Boolean res = result.result();
        if (result.succeeded() && res) {
          completion.complete();
        }
      });
      if (completion.isDone()) {
        break;
      }
    }
    if (!completion.isDone() && started.get()) {
      attemptConnect(completion);
    }
  }

  private void attemptConnect(CompletableAsyncCompletion completion) {
    vertx.setTimer(DELAY, handler -> this.startInternal(completion));
  }

  private void connect(Promise<Boolean> result, URI uri) {
    client
        .websocket(
            uri.getPort(),
            uri.getHost(),
            uri.toString(),
            MultiMap.caseInsensitiveMultiMap().add("origin", "http://localhost"),
            ws -> {
              ws.closeHandler((aVoid) -> attemptConnect(AsyncCompletion.incomplete()));
              ws.exceptionHandler(e -> {
                logger.debug("Error while communicating with ethnetstats", e);

              });
              ws.textMessageHandler(message -> {
                try {
                  JsonNode node = mapper.readTree(message);
                  JsonNode emitEvent = node.get("emit");
                  if (emitEvent.isArray()) {
                    String eventValue = emitEvent.get(0).textValue();
                    if (!result.future().isComplete()) {
                      if (!"ready".equals(eventValue)) {
                        logger.warn(message);
                        result.complete(false);
                      } else {
                        logger.debug("Connected OK! {}", message);
                        result.complete(true);

                        // we are connected and now sending information
                        reportPeriodically(ws);
                        writePing(ws);
                        report(ws);
                      }
                    } else {
                      handleEmitEvent((ArrayNode) emitEvent, ws);
                    }
                  } else {
                    logger.warn(message);
                    result.complete(false);
                  }
                } catch (IOException e) {
                  throw new UncheckedIOException(e);
                }
              });

              writeCommand(ws, "hello", new AuthMessage(nodeInfo, id, secret));
            },
            result::fail);
  }

  private void handleEmitEvent(ArrayNode event, WebSocket ws) {
    String command = event.get(0).textValue();
    switch (command) {
      case "node-pong":
        logger.debug("Received a pong {}", event.get(1));
        if (!waitingOnPong.compareAndSet(true, false)) {
          logger.warn("Received pong when we didn't expect one");
        } else {
          long start = event.get(1).get("clientTime").longValue();
          long latency = (Instant.now().toEpochMilli() - start) / (2 * 1000);
          writeCommand(ws, "latency", "latency", latency);
        }
        break;
      case "history":
        logger.debug("History request {}", event.get(1));
        requestHistory(event.get(1));
        break;
      default:
        logger.warn("Unexpected message {}", command);

    }
  }

  private void requestHistory(JsonNode list) {
    List<UInt256> request = new ArrayList<>();
    for (JsonNode elt : list) {
      request.add(UInt256.fromHexString(elt.asText()));
    }
    historyRequester.accept(request);
  }

  private void writePing(WebSocket ws) {
    waitingOnPong.set(true);
    writeCommand(ws, "node-ping", "clientTime", timeSupplier.get());
  }

  private void reportPeriodically(WebSocket ws) {
    TimeoutStream reportingStream = vertx.periodicStream(REPORTING_PERIOD).handler(ev -> {
      report(ws);
    });
    TimeoutStream pingStream = vertx.periodicStream(PING_PERIOD).handler(ev -> {
      writePing(ws);
    });
    ws.closeHandler(h -> {
      reportingStream.cancel();
      pingStream.cancel();
      attemptConnect(null);
    });
  }

  private void report(WebSocket ws) {
    BlockStats head = newHead.getAndSet(null);
    if (head != null) {
      writeCommand(ws, "block", "block", head);
    }
    Integer count = newTxCount.getAndSet(null);
    if (count != null) {
      writeCommand(ws, "pending", "stats", Collections.singletonMap("pending", count));
    }
    NodeStats nodeStats = newNodeStats.getAndSet(null);
    if (nodeStats != null) {
      writeCommand(ws, "stats", "stats", nodeStats);
    }
    List<BlockStats> newBlocks = newHistory.getAndSet(null);
    if (newBlocks != null && !newBlocks.isEmpty()) {
      writeCommand(ws, "history", "history", newBlocks);
    }
  }

  private void writeCommand(WebSocket ws, String command, Object payload) {
    try {
      String message =
          mapper.writer().writeValueAsString(Collections.singletonMap("emit", Arrays.asList(command, payload)));
      logger.debug("Sending {} message {}", command, message);
      ws.writeTextMessage(message);
    } catch (JsonProcessingException e) {
      throw new UncheckedIOException(e);
    }
  }

  private void writeCommand(WebSocket ws, String command, String key, Object payload) {
    Map<String, Object> body = new HashMap<>();
    body.put("id", id);
    body.put(key, payload);
    writeCommand(ws, command, body);
  }
}

File: ethstats/src/main/kotlin/org/apache/tuweni/ethstats/EthStatsReporter.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.tuweni.ethstats

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ArrayNode
import io.vertx.core.MultiMap
import io.vertx.core.Vertx
import io.vertx.core.http.HttpClient
import io.vertx.core.http.HttpClientOptions
import io.vertx.core.http.WebSocket
import io.vertx.core.http.WebSocketConnectOptions
import io.vertx.kotlin.coroutines.await
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.apache.tuweni.concurrent.AsyncResult
import org.apache.tuweni.concurrent.coroutines.await
import org.apache.tuweni.eth.EthJsonModule
import org.apache.tuweni.units.bigints.UInt256
import org.slf4j.LoggerFactory
import java.io.IOException
import java.io.UncheckedIOException
import java.net.URI
import java.time.Instant
import java.util.Arrays
import java.util.Collections
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference
import kotlin.coroutines.CoroutineContext

/**
 * ETHNetStats reporting service.
 * <p>
 * This service connects to a running ethnetstats service and reports.
 * <p>
 * If the service is not available, the reporter will keep trying to connect periodically. The service will report
 * statistics over time.
 *
 * @param vertx a Vert.x instance, externally managed.
 * @param id the id of the ethstats reporter for communications
 * @param ethstatsServerURIs the URIs to connect to eth-netstats, such as ws://www.ethnetstats.org:3000/api. URIs are
 *        tried in sequence, and the first one to work is used.
 * @param secret the secret to use when we connect to eth-netstats
 * @param name the name of the node to be reported in the UI
 * @param node the node name to be reported in the UI
 * @param port the devp2p port exposed by this node
 * @param network the network id
 * @param protocol the version of the devp2p eth subprotocol, such as eth/63
 * @param os the operating system on which the node runs
 * @param osVer the version of the OS on which the node runs
 * @param historyRequester a hook for ethstats to request block information by number.
 * @param timeSupplier a function supplying time in milliseconds since epoch.
 * @param coroutineContext the coroutine context of the reporter.
 */
class EthStatsReporter(
  private val vertx: Vertx,
  private val id: String,
  private val ethstatsServerURIs: List<URI>,
  private val secret: String,
  private val name: String,
  private val node: String,
  private val port: Int,
  private val network: String,
  private val protocol: String,
  private val os: String,
  private val osVer: String,
  private val historyRequester: (List<UInt256>) -> Unit,
  private val timeSupplier: () -> Instant,
  override val coroutineContext: CoroutineContext = Dispatchers.Default
) : CoroutineScope {

  companion object {
    val mapper = ObjectMapper()

    init {
      mapper.registerModule(EthJsonModule())
    }

    val logger = LoggerFactory.getLogger(EthStatsReporter::class.java)

    const val DELAY = 5000L
    const val REPORTING_PERIOD = 1000L
    const val PING_PERIOD = 15000L
  }

  val nodeInfo = NodeInfo(name, node, port, network, protocol, os = os, osVer = osVer)
  private val started = AtomicBoolean(false)
  private var client: HttpClient? = null
  private val newTxCount = AtomicReference<Int>()
  private val newHead = AtomicReference<BlockStats>()
  private val newNodeStats = AtomicReference<NodeStats>()
  private val newHistory = AtomicReference<List<BlockStats>>()
  private val waitingOnPong = AtomicBoolean(false)

  suspend fun start() {
    if (started.compareAndSet(false, true)) {
      client = vertx.createHttpClient(HttpClientOptions().setLogActivity(true))
      startInternal()
    }
  }

  suspend fun stop() {
    if (started.compareAndSet(true, false)) {
      logger.debug("Stopping the ethstats client service")
      client?.close()
    }
  }

  fun sendNewHead(newBlockStats: BlockStats) {
    newHead.set(newBlockStats)
  }

  fun sendNewPendingTransactionCount(txCount: Int) {
    newTxCount.set(txCount)
  }

  fun sendNewNodeStats(nodeStats: NodeStats) {
    newNodeStats.set(nodeStats)
  }

  fun sendHistoryResponse(blocks: List<BlockStats>) {
    newHistory.set(blocks)
  }

  private suspend fun startInternal() {
    for (uri in ethstatsServerURIs) {
      launch {
        attemptConnect(uri)
      }
    }
  }

  private suspend fun attemptConnect(uri: URI) {
    while (!connect(uri).await() && started.get()) {
      delay(DELAY)
    }
  }

  private suspend fun connect(uri: URI): AsyncResult<Boolean> {
    val result = AsyncResult.incomplete<Boolean>()
    val options = WebSocketConnectOptions().setHost(uri.host).setPort(uri.port)
      .setHeaders(MultiMap.caseInsensitiveMultiMap().add("origin", "http://localhost"))
    val ws = client!!.webSocket(options).await()
    ws.closeHandler { launch { attemptConnect(uri) } }
    ws.exceptionHandler { e ->
      logger.debug("Error while communicating with ethnetstats", e)
    }
    ws.textMessageHandler { message ->
      try {
        val node = mapper.readTree(message)
        val emitEvent = node.get("emit")
        if (emitEvent.isArray) {
          val eventValue = emitEvent.get(0).textValue()
          if (!result.isDone) {
            if ("ready" != eventValue) {
              logger.warn(message)
              result.complete(false)
            } else {
              logger.debug("Connected OK! {}", message)
              result.complete(true)

              // we are connected and now sending information
              reportPeriodically(uri, ws)
              writePing(ws)
              report(ws)
            }
          } else {
            handleEmitEvent(emitEvent as ArrayNode, ws)
          }
        } else {
          logger.warn(message)
          result.complete(false)
        }
      } catch (e: IOException) {
        throw UncheckedIOException(e)
      }
    }

    writeCommand(ws, "hello", AuthMessage(nodeInfo, id, secret))
    return result
  }

  private fun handleEmitEvent(event: ArrayNode, ws: WebSocket) {
    val command = event.get(0).textValue()
    when (command) {
      ("node-pong") -> {
        logger.debug("Received a pong {}", event.get(1))
        if (!waitingOnPong.compareAndSet(true, false)) {
          logger.warn("Received pong when we didn't expect one")
        } else {
          val start = event.get(1).get("clientTime").longValue()
          val latency = (Instant.now().toEpochMilli() - start) / (2 * 1000)
          writeCommand(ws, "latency", "latency", latency)
        }
      }
      "history" -> {
        logger.debug("History request {}", event.get(1))
        requestHistory(event.get(1))
      }
      else ->
        logger.warn("Unexpected message {}", command)
    }
  }

  private fun requestHistory(list: JsonNode) {
    val request = mutableListOf<UInt256>()
    for (elt in list) {
      request.add(UInt256.fromHexString(elt.asText()))
    }
    historyRequester(request)
  }

  private fun writePing(ws: WebSocket) {
    waitingOnPong.set(true)
    writeCommand(ws, "node-ping", "clientTime", timeSupplier())
  }

  private fun reportPeriodically(uri: URI, ws: WebSocket) {
    val reportingStream = vertx.periodicStream(REPORTING_PERIOD).handler {
      report(ws)
    }
    val pingStream = vertx.periodicStream(PING_PERIOD).handler {
      writePing(ws)
    }
    ws.closeHandler {
      reportingStream.cancel()
      pingStream.cancel()
      launch {
        attemptConnect(uri)
      }
    }
  }

  private fun report(ws: WebSocket) {
    val head = newHead.getAndSet(null)
    if (head != null) {
      writeCommand(ws, "block", "block", head)
    }
    val count = newTxCount.getAndSet(null)
    if (count != null) {
      writeCommand(ws, "pending", "stats", Collections.singletonMap("pending", count))
    }
    val nodeStats = newNodeStats.getAndSet(null)
    if (nodeStats != null) {
      writeCommand(ws, "stats", "stats", nodeStats)
    }
    val newBlocks = newHistory.getAndSet(null)
    if (newBlocks != null && !newBlocks.isEmpty()) {
      writeCommand(ws, "history", "history", newBlocks)
    }
  }

  private fun writeCommand(ws: WebSocket, command: String, payload: Any) {
    try {
      val message =
        mapper.writer().writeValueAsString(Collections.singletonMap("emit", Arrays.asList(command, payload)))
      logger.debug("Sending {} message {}", command, message)
      ws.writeTextMessage(message)
    } catch (e: JsonProcessingException) {
      throw UncheckedIOException(e)
    }
  }

  private fun writeCommand(ws: WebSocket, command: String, key: String, payload: Any) {
    val body = mutableMapOf<String, Any>()
    body["id"] = id
    body[key] = payload
    writeCommand(ws, command, body)
  }
}
